internal/controller/etcdopstask/reconcile_task.go

```diff
@@ -0,0 +1,222 @@
+)
+
+const (
+	// ErrDuplicateTask represents the error in case of a duplicate task for the same etcd.
+	ErrDuplicateTask druidv1alpha1.ErrorCode = "ERR_DUPLICATE_TASK"
+)
+
+// reconcileTask manages the lifecycle of an EtcdOpsTask resource.
+// It executes a series of step functions to ensure the task is processed correctly.
+func (r *Reconciler) reconcileTask(ctx context.Context, logger logr.Logger, taskObjKey client.ObjectKey, taskHandler handler.Handler) ctrlutils.ReconcileStepResult {
+	logger = logger.WithValues("op", "reconcileTask")
+	logger.Info("Triggering task execution flow")
+
+	reconcileStepFns := []reconcileFn{
+		r.ensureTaskFinalizer,
+		r.transitionToPendingState,
+		r.admitTask,
+		r.transitionToInProgressState,
+		r.runTask,
+	}
+
+	for _, fn := range reconcileStepFns {
+		if result := fn(ctx, logger, taskObjKey, taskHandler); ctrlutils.ShortCircuitReconcileFlow(result) {
+			return result
+		}
+	}
+	logger.Info("Task execution flow completed")
+	return ctrlutils.ContinueReconcile()
+}
+
+// ensureTaskFinalizer checks if the EtcdOpsTask has the finalizer.
+// If not, it adds the finalizer and updates the task status.
+func (r *Reconciler) ensureTaskFinalizer(ctx context.Context, _ logr.Logger, taskObjKey client.ObjectKey, _ handler.Handler) ctrlutils.ReconcileStepResult {
+	task, err := r.getTask(ctx, taskObjKey)
+	if err != nil {
+		return ctrlutils.ReconcileWithError(err)
+	}
+
+	if controllerutil.ContainsFinalizer(task, druidapicommon.EtcdOpsTaskFinalizerName) {
+		return ctrlutils.ContinueReconcile()
+	}
+
+	if err := kubernetes.AddFinalizers(ctx, r.client, task, druidapicommon.EtcdOpsTaskFinalizerName); err != nil {
+		if updateErr := r.updateTaskStatus(ctx, task, taskStatusUpdate{
+			Operation: &druidv1alpha1.LastOperation{
+				Type:        druidv1alpha1.LastOperationTypeReconcile,
+				State:       druidv1alpha1.LastOperationStateRequeue,
+				Description: fmt.Sprintf("ensureTaskFinalizer failed to add finalizer: %v", err),
+			},
+		}); updateErr != nil {
+			return ctrlutils.ReconcileWithError(errors.Wrapf(err, "failed to update status after error: %v", updateErr))
+		}
+		return ctrlutils.ReconcileWithError(err)
+	}
+	return ctrlutils.ContinueReconcile()
+}
+
+// transitionToPendingState sets the task.status.state to Pending if not already set.
+func (r *Reconciler) transitionToPendingState(ctx context.Context, logger logr.Logger, taskObjKey client.ObjectKey, _ handler.Handler) ctrlutils.ReconcileStepResult {
+	logger = logger.WithValues("step", "transitionToPendingState")
+	logger.Info("Executing step: transitionToPendingState")
+
+	task, err := r.getTask(ctx, taskObjKey)
+	if err != nil {
+		return ctrlutils.ReconcileWithError(err)
+	}
+
+	if task.Status.State != nil {
+		return ctrlutils.ContinueReconcile()
+	}
+	// Set the task state to Pending
+	// and update the LastTransitionTime.
+	pendingState := druidv1alpha1.TaskStatePending
+	if err := r.updateTaskStatus(ctx, task, taskStatusUpdate{
```

Comment:
It is not semantically correct to just update the task state. AFAIU - Phase is high level and TaskState is representing where in the phase the task processing has reached. From that perspective `Pending` TaskState cannot exist without belonging to a Phase. Thus the Phase should always be `Admit` for TaskState `Pending` indicating clearly that the admit of the task is pending.